src/libraries/System.Net.Ping/src/System/Net/NetworkInformation/Ping.PingUtility.cs

```diff
@@ -73,33 +73,41 @@ private PingReply SendWithPingUtility(IPAddress address, byte[] buffer, int time
 
         private async Task<PingReply> SendWithPingUtilityAsync(IPAddress address, byte[] buffer, int timeout, PingOptions? options)
         {
-            using (Process p = GetPingProcess(address, buffer, timeout, options))
-            {
-                var processCompletion = new TaskCompletionSource(TaskCreationOptions.RunContinuationsAsynchronously);
-                p.EnableRaisingEvents = true;
-                p.Exited += (s, e) => processCompletion.SetResult();
-                p.Start();
+            CancellationToken timeoutOrCancellationToken = _timeoutOrCancellationSource!.Token;
 
+            using Process pingProcess = GetPingProcess(address, buffer, timeout, options);
+            pingProcess.Start();
+
+            var timedOutOrCanceled = false;
+            try
+            {
+                string stdout;
                 try
                 {
-                    await processCompletion.Task.WaitAsync(TimeSpan.FromMilliseconds(timeout)).ConfigureAwait(false);
+                    await pingProcess.WaitForExitAsync(timeoutOrCancellationToken).ConfigureAwait(false);
+
+                    stdout = await pingProcess.StandardOutput.ReadToEndAsync(timeoutOrCancellationToken).ConfigureAwait(false);
                 }
-                catch (TimeoutException)
+                catch when (timeoutOrCancellationToken.IsCancellationRequested)
```

Comment:
We don't want to only catch OperationCanceledException?  If the wait failed with an unrelated exception at appx the same time that cancellation was requested, we'll end up eating that exception and treating it as a timeout?

Reply:
> Were you imagining

The latter.

> In that case, does it matter? We have a race condition where multiple things happened that should have aborted the ping and we are correctly informing the caller about one of those things.

When this situation typically arises elsewhere,  we make the same argument,  but in those cases there's the added complication that the cancellation could have actually caused the other exception,  and then it's actually preferable to show the cancelation one because the other is fake in a sense.  Here,  though, I don't think that's the case, in which case we generally want to prioritize real failures over manufactured ones from timeouts and cancellation,  as they typically convey better information about the state of the system.